newlib/libc/sys/vita/fcntl.c

```diff
@@ -114,7 +113,24 @@ int _fcntl_r(struct _reent *reent, int fd, int cmd, ...)
 		}
 	}
 
+	if (fdmap->type == VITA_DESCRIPTOR_FILE)
+	{
+		if (cmd == F_GETFL)
+		{
+			int flags = fdmap->flags;
+			__vita_fd_drop(fdmap);
+			return flags;
+		}
+
+		if (cmd == F_SETFL)
+		{
+			fdmap->flags = arg;
+			__vita_fd_drop(fdmap);
+			return 0;
```

Comment:
hm.... i see....

very first time, i think it simply only allows same flags
```
if ((fdmap->flags & arg) == arg) {
  __vita_fd_drop(fdmap);
  return 0;
}
```
btw, `fdopen_r` [implementation](https://github.com/vitasdk/newlib/blob/b89e5bc1/newlib/libc/stdio/fdopen.c#L91-L94) automatically appends `O_APPEND` into new flag.

so it should be...

```
int new_flag = arg;
// original doesn't have O_APPEND, and input has O_APPEND,
// it can be possible to be added by fdopen_r logic.
// drop this flag from input.
if (!(fdmap->flags & O_APPEND) && (arg & O_APPEND)) {
  arg &= ~O_APPEND;
}
// our fcntl cannot update actual flags,
// that this reason, we only check arg contains same flags
if ((fdmap->flags & arg) == arg) {
  __vita_fd_drop(fdmap);
  return 0
}
```
i'm not sure that this is correct or not.
maybe just checking same flag value (`fdmap->flags == arg`) is batter.

ps. imo, we might have to reimplement fdopen too...
for example, both implementation (pr version and my suggestions) have problem.
F_SETFL doesn't affect actual flags of file description of the vita system, it wouldn't work programmer's expectation in file IO

so, more good way is use [fdopen of the sdk version](https://github.com/vitasdk/vita-headers/blob/9ec49695/db/360/SceLibc.yml#L80) or newlib's fdopen always returns an error when fd is file descriptor.

Reply:
fdopen not only changes flags, but also sets them on internal file descriptor, see
https://github.com/vitasdk/newlib/blob/vita/newlib/libc/stdio/fdopen.c#L86
but those are only used for fwrite, not for write.
I've pushed another fix that
1) updates those flags in fcntl (so fwrite works as intended)
2) write now handles O_APPEND flags by seeking to the end of file.

This simple test
```
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#if defined(__vita__)
#include <psp2/kernel/clib.h>
#define printf sceClibPrintf
#define FPATH "ux0:/data/fd_share.txt"
#else
#define FPATH "fd_share.txt"
#endif

int main() {
    int fd = open(FPATH, O_RDWR | O_CREAT, 0644);

    if (fd < 0)
    {
        printf("errno: 0x%08x\n", errno);
        return 0;
    }

    // append to empty file
    write(fd, "aaaaaaaaaa", 10);

    off_t cur = lseek(fd, 1, SEEK_SET);
    printf("offset after being set to 1: %ld\n", (long)cur);

    // append
    write(fd, "bbbbbbbb", 8);

    cur = lseek(fd, 0, SEEK_CUR);
    printf("offset after appending bbbbbbbb: %ld\n", (long)cur);

    cur = lseek(fd, 2, SEEK_SET);
    printf("offset after being set to 2: %ld\n", (long)cur);

    // now toggle "append mode"
    int open_flag = fcntl(fd, F_GETFL);
    if (fcntl(fd, F_SETFL, open_flag | O_APPEND) == -1) {
        printf("failed to set flag\n");
        return 0;
    }

    cur = lseek(fd, 0, SEEK_CUR);
    printf("offset after setting O_APPEND: %ld\n", (long)cur);

    // write without appending
    write(fd, "cccc", 4);

    cur = lseek(fd, 0, SEEK_CUR);
    printf("offset after writing cccc: %ld\n", (long)cur);


    close(fd);

    return 0;
}
```
produces same results on PC and vita now.

Reply:
sorry. i'm late.
i want to make a reply comment after testing in linux, but really sorry i cannot made time for it.
so, I make a comment first, then test it asap...

my review means, if someone open a file with `readonly`, it's not able to SETFL with `readwrite`.

for example;
```
int fd = open(FILE_PATH, O_RDONLY);
...
fcntl(fd, FSETFL, O_RDWR); // <- if it work well, system support this in os level or filesystem level
```

imo it wouldn't work well. that because newlib doesn't make/handle internal filesystem level. it just borrow `vitasdk`'s wrapping function.

that this reason, i couldn't believe this patch would work correctly.

just bypassing the error cannot be a goal of the libc.
if we have a case which we aren't able to handling, we have to block to access this until to find solution.
this is reason why a call return an error with `ENOTSUP`

Reply:
in my test on linux, fcntl was pass, but process cannot write file.

```
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>

int main() {
    int fd = open("test.txt", O_RDONLY);
    printf("fd: %d\n", fd);

    int fl = fcntl(fd, F_GETFL);
    printf("fl: %d\n", fl);
    int ret = fcntl(fd, F_SETFL, fl | O_RDWR);
    printf("ret: %d\n", ret);

    if (ret < 0) {
        printf("errno: %d\n", errno);
        goto exit;
    }
    printf("fl: %d", fcntl(fd, F_GETFL));

    write(fd, "asdf", 4);
exit:
    close(fd);
}
```
idk why it's callable in linux, but SETFL could happen nothing.

```
fd: 3
fl: 32768
ret: 0
fl: 32768
```

Reply:
ok, [in linux manpage](https://man7.org/linux/man-pages/man2/fcntl.2.html), they said file flags are ignored in linux

```
              Set the file status flags to the value specified by arg.
              File access mode (O_RDONLY, O_WRONLY, O_RDWR) and file
              creation flags (i.e., O_CREAT, O_EXCL, O_NOCTTY, O_TRUNC)
              in arg are ignored.  On Linux, this operation can change
              only the O_APPEND, O_ASYNC, O_DIRECT, O_NOATIME, and
              O_NONBLOCK flags.  It is not possible to change the
              O_DSYNC and O_SYNC flags; see BUGS, below.
```

[opengroup spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fcntl.html) doesn't have similar info.

```
Set the file descriptor flags defined in <fcntl.h>, that are associated with fildes, to the third argument, arg, taken as type int.
If the FD_CLOEXEC flag in the third argument is 0, the file descriptor shall remain open across the exec functions;
otherwise, the file descriptor shall be closed upon successful execution of one of the exec functions.
```
so that can be linux specific behavior.

[in freebsd case](https://man.freebsd.org/cgi/man.cgi?query=fcntl&sektion=2&n=1), they also just print
`       F_SETFL		 Set descriptor	status flags to	arg.`

but they are also listed up possible flag list. (not exists file operation flags in here).

so we have two options.
1. follow freebsd options; but patch have to more bigger. so i suggest that do it later.
2. raise just an error. imo it still have compatibility to opengroup spec (but some libs can be break..)

Reply:
You are reading old openspec doc.
```
Bits corresponding to the file access mode and the file creation flags, as defined in <fcntl.h>, that are set in arg shall be ignored
```
https://pubs.opengroup.org/onlinepubs/9799919799/functions/fcntl.html

Reply:
Freebsd also lists available flags:
https://man.freebsd.org/cgi/man.cgi?query=fcntl&sektion=2&n=1

```
The flags for the F_GETFL and F_SETFL commands are as follows:

       O_NONBLOCK   Non-blocking  I/O;	if  no	data is	available to a read(2)
		    system call, or if a write(2) operation would  block,  the
		    read or write call returns -1 with the error EAGAIN.

       O_APPEND	    Force each write to	append at the end of file; corresponds
		    to the O_APPEND flag of open(2).

       O_DIRECT	    Minimize  or  eliminate  the  cache	effects	of reading and
		    writing.  The system will attempt  to  avoid  caching  the
		    data  you  read  or	write.	If it cannot avoid caching the
		    data, it will minimize the impact  the  data  has  on  the
		    cache.   Use  of  this flag	can drastically	reduce perfor-
		    mance if not used with care.

       O_ASYNC	    Enable the SIGIO signal to be sent to  the	process	 group
		    when  I/O  is possible, e.g., upon availability of data to
		    be read.

       O_SYNC	    Enable synchronous writes.	Corresponds to the O_SYNC flag
		    of open(2).	 O_FSYNC is an historical synonym for O_SYNC.

       O_DSYNC	    Enable  synchronous	 data  writes.	 Corresponds  to   the
		    O_DSYNC flag of open(2).
```

Reply:
ah, I copied wrong part of the opengroups document.

1. file access mode and creation mode should be ignored.
2. if system doesn't support async operation, they doesn't defined when receive `O_NONBLOCK`

anyway we have to filter first,

```
#define SUPPORT_SETFL_FLAGS (O_NONBLOCK | O_APPEND | O_DIRECT | O_ASYNC | O_SYNC | O_DSYNC)
new_flag = (input_flag & SUPPORT_SETFL_FLAGS | old_flags);
```
then have to implement each flags
```
ret = ...
if (new_flag & O_NONBLOCK) {
  ...
  ret |= O_NONBLOCK;
}
...
fd.flag = ret
return ret;
```

🤔

Reply:
I agree with filtering (i pushed a change to do that), but i don't think we have to implement each flag (we only support O_APPEND anyway)

Reply:
posix standard supports only
```
O_APPEND
Set append mode.
O_NONBLOCK
Non-blocking mode.
O_SYNC
Write according to synchronized I/O file integrity completion.
```
O_SYNC is the default behavior for sceIoWrite, 
We support O_APPEND
And for O_NONBLOCK:
```
If fildes does not support non-blocking operations, it is unspecified whether the O_NONBLOCK flag will be ignored.
```
so we can just ignore it for now.

Reply:
yes i agree that. and ty update pr patches.

can you check last my comment about latest patchset??